### ISSUE TITLE
Image size

### ISSUE THREAD

**Issue by exelban** [RyanZotti/Self-Driving-Car#12]
Did you try to resize input image to 128x128 or smaller?
Because now you are working with so big images, and it takes too much time to train nn.
And what FPS you receive from raspberry pi on computer?

**Comment by RyanZotti** [comment]
I haven't tried that, but that's a good idea. I've tried max pooling, which effectively shrinks each dimension by a factor of two, but it didn't seem to improve performance much. I'm training on a GPU, so training is relatively fast. 

During data collection I was able to capture about 20 frames per second from the ffmpeg utility on the Pi. However, during live tests after model deployment, I was only able to score 2-3 frames per second. It took between 300-400 milliseconds to score each frame on the Pi, which is way to slow. That's why I chose to score the frames on my laptop, even though that added some network latency (about 40 milliseconds).

**Comment by exelban** [comment]
@RyanZotti I’m working on the same project.
I’m training network on GPU too, but you use too big image size. Try 128x128 or smaller.

Next problem is data collection. Best off all its run webserver on rpi, stream video to this web page throw ffmpeg. And connect to rpi via VNC. I receive > 30fps.
Ffmpeg is not the best solution better use PiCamera.

**Comment by RyanZotti** [comment]
@exelban I agree. The Raspberry Pi camera was my first choice, but I kept ruining my cameras, probably because I have too many exposed wires in my car and accidentally shorted something. The web cam I use has a protective plastic coating around the camera wires so the camera is much more durable. When I went with the web cam I had to use ffmpeg, although I tried alternatives. Ffmpeg was painful to set up.

It's great to hear that you're working on a similar project. What's  your target variable? Do you use arrow keys (classification) or something more granular likes turn radius (regression)?  

I'm still trying to improve the accuracy of my convnet, but I'm getting poor results. My best guess is that I don't have enough training records. I only have about 30,000 frames, which is small. I'm testing each of the latest techniques (dropout, 3d convolution, batch normalization, etc) to see if I can improve the performance. So far the simple fully connected neural net is the best performing model, but in theory the convent should be much better. Have you faced a similar convnet accuracy problem?

**Comment by exelban** [comment]
@RyanZotti i use rpi camera with fish-eye lens. It give me full view of "road".
I run few processes on rpi:
- car controller;
- web server for controlling car and writing data.

I simply detect key pressing (arrows) and write it to frame when key is pressed.I have rc car with servo drive for steering so i use arduino to control.

For training i use few hour of drive on 30fps, so usually its 540000 frames with data. Also i create simple system based on opencv for detection lines on road (canny edge detector and hough transform). This system gives me opportunity to estimate nn work.

With the model i have problems, because it takes so much times to train nn. So i can try only few models per month (in free time). As you now each data need own model. But best results i receive with fractional Max-Pooling. But i think that system like this need some other model (not only cnn). Also try to preprocessing image before push to nn. For example use canny egde detector.
I have tried all techniques like dropout, 3d convolution, batch normalization, etc. And combine them, but dont receive good accuracy.

On last tests i receive acurracy ~70%, but i update all code to python 3.5 and cant run model on rpi (have some problems with compiling tensorflow on pri with python 3.5).

**Comment by exelban** [comment]
@RyanZotti also try to change road. You can just buy few A1 and draw dark road lines. Its will be better to recognise the road for nn.

**Comment by RyanZotti** [comment]
@exelban That's very interesting. I'm also stuck at ~70% accuracy. Do you think that could be due to the fact that we're using arrow keys instead of something more granular like angular displacement? After inspecting my data, I see a lot of cases where my car could arguably go either way, for example, left or straight on a wide stretch of road where a more precise measurement might have better captured that nuance.

Do you use canny edge detection to do thresholding to make a binary (black/white) representation of your images? I haven't tried that yet, but it sounds promising. Although I was hoping that the state-of-the-art neural net techniques like convnets would make that unnecessary. 

Nvidia made a real self-driving car and wrote a [paper](https://images.nvidia.com/content/tegra/automotive/images/2016/solutions/pdf/end-to-end-dl-using-px.pdf) on it. Not sure if you've read it. In it they say:

> We then sample that video at 10 FPS. A higher sampling rate would result in including images that are highly similar and thus not provide much useful information.

That could reduce the size of your dataset. I have a lot of images that are very similar. I haven't performed sampling only because I have the opposite problem -- too few images. 

Do you use a circular track? I don't have enough space in my apartment to make a circular track, and so it's very time consuming to generate lots of data because I have to pick up my car and put it at the beginning each time whereas a circular (or closed loop) track I could keep repeating the course continuously until I run out of battery.

**Comment by RyanZotti** [comment]
Are you absolutely sure you need python 3.5? I was able to get Tensorflow working on the Pi with ease by pulling the binaries following the instructions [here](https://github.com/samjabrahams/tensorflow-on-raspberry-pi/). The instructions use python 3.3 though. Even if you get it working, you'll probably arrive at the same conclusion I did -- the Pi is far too slow for making predictions in real time, at least for anything but very slow speeds. 300-400 milliseconds per frame (what I got) is extremely slow. My car kept going off course because it would see frames when it was too late.

**Comment by exelban** [comment]
@RyanZotti i read this paper, but this model gives me only ~50%.
I have tried different controls. But also i dont set "full power" to drive when key is pressed. I have function witch increase value from minimum to max in time. It's give me opportunity to better control and better data for training. 

And this paper is only theory, i dont find any examples. Maybe this model work in their simulation but in my case it does not working. Better read [DriveWorks SDK](http://www.nvidia.com/object/driveworks.html) documentation (you must to have access to this program on nvidia site). 

Its because python 3.5 have implemented many function which i use (also i dont want to rewrite all code). I use this repo for compiling, but bazel have some bugs and cant compile.

**Comment by exelban** [comment]
@RyanZotti i write for my degree work system for recognition road lines and following approach them on opencv. So i have car, road, rpi and in free time i collecting data and trying to make similar system with nn for myself (and for github of course)).

**Comment by RyanZotti** [comment]
@exelban Ok, I'll take a look at DriveWorks. I never saw that before. 

Have you posted your code on GitHub? I couldn't find anything on your GitHub profile. 

The 50% to 70% accuracy jump is huge. Did that increase mostly come from the canny edge detector? How many color channels do you use for your image? Just one for black and white?

**Comment by exelban** [comment]
@RyanZotti if you want i can send you documentation for DriveWork (i hope that nvidia does not ban me for that) on email. Because their check submits for a few month(
A want to post all code, but dont have enough time to clean up all and write docs.

I dont make tests with preprocessing image, now in free time i try to improve opencv code( But in work on this i have some ideas for nn.

I working with 3 channels, but have plans to test only 1 (black and white).

I install tensorflow for python 3.5 and make quick tests on cifar10. And send the results.

**Comment by RyanZotti** [comment]
I don't want you to get in trouble for sending me unauthorized DriveWork documents. I'll try to find some other way to access the docs. Until just a few minutes ago I hadn't read the text at the bottom that says: 

> NVIDIA DriveWorks is available to select Automakers, Tier 1 Suppliers, and Research Institutions working on developing systems that enable cars to drive themselves. - See more at: http://www.nvidia.com/object/driveworks.html#sthash.PWV7oeAk.dpuf

Which research institution are you part of? Is that how you got access to the SDK?

If you could replicate the canny edge detector preprocessing for your self driving car on the cifar10 dataset and post your code on GitHub that would be great. I'm very curious how you've done that.

**Comment by exelban** [comment]
@RyanZotti i just write my university title. And that im writing my degree work and want to read how DriveWork work.

I have cifar10 [repo ](https://github.com/exelban/tensorflow-cifar-10) (its not for self driving, its just object recognition dataset), max i receive 76% on this model. But its not better result. I just use this nn for testing performance on different system.

**Comment by exelban** [comment]
@RyanZotti receive next results on Raspberry Pi 3:
- 1 image from cifar10 takes 310ms;
- 30 images from cifar10 takes 4.12s;
- 30 images from cifar10 takes 7.68s.

The same on gtx 960 4GB:
- 1 image from cifar10 takes 650ms;
- 30 images from cifar10 takes 960ms;
- 30 images from cifar10 takes 1.14s.

And i dont understand why one image quickly process on RPi than on gtx 960.
Tomorrow will make more tests.

**Comment by RyanZotti** [comment]
@exelban That's comparable to what I got on the Pi.

I've never tested GPU latency, but those results make sense given that GPUs are built for batch parallelization (i.e., throughput) and not lead time like CPUs. I've only heard of people using GPUs for training (batch) not for making predictions in real time. Anyways, your laptop/desktop's CPU should be able to process a single image in 20-30 ms or less (i.e., much faster than both the Pi and GPU). That's why I chose my laptop over the Pi for the real-time predictions. The GPU is 14x faster than my laptop for training though because that's where throughput matters. 

In an earlier comment when you said:

> On last tests i receive acurracy ~70%, but i update all code to python 3.5 and cant run model on rpi (have some problems with compiling tensorflow on pri with python 3.5).

Did you mean that you got 70% accuracy on your car's dataset or on cifar10?

**Comment by exelban** [comment]
@RyanZotti car's dataset

**Comment by exelban** [comment]
@RyanZotti run alexnet model on raspberry with trained data (3k iteration). Without any preprocessing.
So i receive next results:
- [INFO] elasped time: 19.86
- [INFO] approx. FPS: 5.03

I think then after adding preprocessing i will lose 100-200ms on each frame, but nn will be faster too. Because image will be smaller and model will be have less variables.

**Comment by RyanZotti** [comment]
@exelban The numbers between 60000 - 120000 represent your steering target/response variable? Do those numbers represent degrees or something else? For example, does 60000 correspond to the most extreme left turn and 120000 the most extreme right turn? I'm not sure how to interpret those numbers. They also seem to be really large. 

If I understand your second comment correctly, AlexNet has around 200ms latency per frame (1000 ms / 5 FPS) on your Raspberry Pi. Is my understanding correct?

**Comment by exelban** [comment]
@RyanZotti yes, its represent steering (left - 120000, right - 60000).
I make repo with collecting data from rpi. You just need run 2 python scripts on rpi, and start web server on laptop. Receive 18-25 fps. App collecting frames and make .txt file with control data.
If you want to try, i will open repo. Because now i make secret (dont have time to make readme).

**Comment by RyanZotti** [comment]
@exelban No rush. I'm just working on optimizing my neural networks at the moment. I probably won't make any changes to how I collect data or label left and right turns, since that would require a major overhaul of my code.

**Comment by exelban** [comment]
@RyanZotti ok, now i dont have time for playing with car. At first i want to change car model, because now i have 1:10 and cant drive in room(

If you will be interested: https://github.com/exelban/raspberry-pi-camera-server.
But you must to change file drive.py and write your code for steering car.
Also in app.js line 47 and 60, 64 change parameters witch calculate for me value from 0 to 70 to my value.

Now you can control car throw joystick in browser or arrow keys.

**Comment by RyanZotti** [comment]
@exelban Using the joystick sounds like a good idea. That's probably a lot better than using arrow keys. I'll take a closer look at your code once I've finished all of my neural network optimizations.

**Comment by AlickCameron** [comment]
Good discussion.

**Comment by RyanZotti** [comment]
@exelban I finally figured out what was wrong with the deep convnet. The deep convnet is now by far my best model. I had to replace:

    cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y_conv), reduction_indices=[1]))

with:

    cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=pred, labels=y_))

You can see more of an explanation [here](https://github.com/tensorflow/tensorflow/issues/2462).
